File: aa-tree-set/AaTreeSet.java
/* 
 * AA tree set (Java)
 * 
 * Copyright (c) 2018 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/aa-tree-set
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

import java.util.AbstractSet;
import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Set;
import java.util.Stack;


public final class AaTreeSet<E extends Comparable<? super E>> extends AbstractSet<E> {
	
	/*---- Fields ----*/
	
	private Node<E> root;  // Never null
	
	private int size;
	
	
	
	/*---- Constructors ----*/
	
	public AaTreeSet() {
		clear();
	}
	
	
	public AaTreeSet(Collection<? extends E> coll) {
		this();
		Objects.requireNonNull(coll);
		addAll(coll);
	}
	
	
	
	/*---- Methods ----*/
	
	@SuppressWarnings("unchecked")
	public void clear() {
		root = (Node<E>)Node.EMPTY_LEAF;
		size = 0;
	}
	
	
	public int size() {
		return size;
	}
	
	
	public boolean contains(E val) {
		Objects.requireNonNull(val);
		for (Node<E> node = root; node != Node.EMPTY_LEAF; ) {
			int cmp = val.compareTo(node.value);
			if (cmp < 0)
				node = node.left;
			else if (cmp > 0)
				node = node.right;
			else
				return true;
		}
		return false;
	}
	
	
	public boolean add(E val) {
		Objects.requireNonNull(val);
		if (size == Integer.MAX_VALUE)
			throw new IllegalStateException("Maximum size reached");
		if (contains(val))
			return false;
		root = root.add(val);
		size++;
		return true;
	}
	
	
	public boolean remove(E val) {
		Objects.requireNonNull(val);
		if (!contains(val))
			return false;
		root = root.remove(val);
		size--;
		return true;
	}
	
	
	public Iterator<E> iterator() {
		return new Iter();
	}
	
	
	void checkStructure() {
		HashSet<Node<E>> visited = new HashSet<Node<E>>();
		if (root.checkStructure(visited) != size || visited.size() != size)
			throw new AssertionError();
	}
	
	
	
	/*---- Helper class: AA tree node ----*/
	
	private static final class Node<E extends Comparable<? super E>> {
		
		/*-- Fields --*/
		
		public E value;
		private byte level;
		public Node<E> left;
		public Node<E> right;
		
		
		/*-- Constructors --*/
		
		public Node() {
			value = null;
			level = 0;
			left  = null;
			right = null;
		}
		
		
		@SuppressWarnings("unchecked")
		public Node(E val) {
			assert val != null;
			value = val;
			level = 1;
			left  = (Node<E>)EMPTY_LEAF;
			right = (Node<E>)EMPTY_LEAF;
		}
		
		
		/*-- Methods --*/
		
		public Node<E> add(E val) {
			if (this == EMPTY_LEAF)
				return new Node<>(val);
			int cmp = val.compareTo(value);
			if (cmp < 0)
				left = left.add(val);
			else if (cmp > 0)
				right = right.add(val);
			else
				throw new AssertionError("Value already in tree");
			return skew().split();  // Rebalance this node
		}
		
		
		@SuppressWarnings("unchecked")
		public Node<E> remove(E val) {
			if (this == EMPTY_LEAF)
				throw new AssertionError("Value not in tree");
			int cmp = val.compareTo(value);
			if (cmp < 0)
				left = left.remove(val);
			else if (cmp > 0)
				right = right.remove(val);
			else {  // Remove value at this node
				if (left != EMPTY_LEAF) {
					// Find predecessor node
					Node<E> temp = left;
					while (temp.right != EMPTY_LEAF)
						temp = temp.right;
					value = temp.value;  // Replace value with predecessor
					left = left.remove(value);  // Remove predecessor node
				} else if (right != EMPTY_LEAF) {
					// Find successor node
					Node<E> temp = right;
					while (temp.left != EMPTY_LEAF)
						temp = temp.left;
					value = temp.value;  // Replace value with successor
					right = right.remove(value);  // Remove successor node
				} else {
					assert level == 1;
					return EMPTY_LEAF;
				}
			}
			
			// Rebalance this node if a child was lowered
			if (this.level == Math.min(left.level, right.level) + 1)
				return this;
			if (right.level == this.level)
				right.level--;
			this.level--;
			Node<E> result = this.skew();
			result.right = result.right.skew();
			if (result.right.right != EMPTY_LEAF)
				result.right.right = result.right.right.skew();
			result = result.split();
			result.right = result.right.split();
			return result;
		}
		
		
		/* 
		 *       |          |
		 *   A - B    ->    A - B
		 *  /   / \        /   / \
		 * 0   1   2      0   1   2
		 */
		private Node<E> skew() {
			assert this != EMPTY_LEAF;
			if (left.level < this.level)
				return this;
			Node<E> result = this.left;
			this.left = result.right;
			result.right = this;
			return result;
		}
		
		
		/* 
		 *   |                      |
		 *   |                    - B -
		 *   |                   /     \
		 *   A - B - C    ->    A       C
		 *  /   /   / \        / \     / \
		 * 0   1   2   3      0   1   2   3
		 */
		private Node<E> split() {
			assert this != EMPTY_LEAF;
			// Must short-circuit because if right.level < self.level, then right.right might be null
			if (right.level < this.level || right.right.level < this.level)
				return this;
			Node<E> result = right;
			this.right = result.left;
			result.left = this;
			result.level++;
			return result;
		}
		
		
		int checkStructure(Set<Node<E>> visitedNodes) {
			if (this == EMPTY_LEAF)
				return 0;
			if (!visitedNodes.add(this))
				throw new AssertionError();
			
			if (value == null || left == null || right == null)
				throw new AssertionError();
			if (!(this.level > 0 && this.level == left.level + 1 && (this.level == right.level + 1 || this.level == right.level)))
				throw new AssertionError();
			if (this.level == right.level && this.level == right.right.level)  // Must short-circuit evaluate
				throw new AssertionError();
			if (left != EMPTY_LEAF && !(left.value.compareTo(this.value) < 0))
				throw new AssertionError();
			if (right != EMPTY_LEAF && !(right.value.compareTo(this.value) > 0))
				throw new AssertionError();
			
			int size = 1 + left.checkStructure(visitedNodes) + right.checkStructure(visitedNodes);
			if (size < (1 << level) - 1)
				throw new AssertionError();
			// Not checked, but (size <= 3^level - 1) is also true
			return size;
		}
		
		
		@SuppressWarnings("rawtypes")
		public static final Node EMPTY_LEAF = new Node();
		
	}
	
	
	
	/*---- Helper class: Binary search tree iterator ----*/
	
	// Note: Not fail-fast on concurrent modification.
	private final class Iter implements Iterator<E> {
		
		/*-- Fields --*/
		
		private Stack<Node<E>> stack;
		
		
		/*-- Constructors --*/
		
		public Iter() {
			stack = new Stack<>();
			for (Node<E> node = root; node != Node.EMPTY_LEAF; node = node.left)
				stack.push(node);
		}
		
		
		/*-- Methods --*/
		
		public boolean hasNext() {
			return !stack.isEmpty();
		}
		
		
		public E next() {
			if (!hasNext())
				throw new NoSuchElementException();
			Node<E> node = stack.pop();
			E result = node.value;
			for (node = node.right; node != Node.EMPTY_LEAF; node = node.left)
				stack.push(node);
			return result;
		}
		
		
		public void remove() {
			throw new UnsupportedOperationException();
		}
		
	}
	
}

File: aa-tree-set/aatreeset.py
# 
# AA tree set (Python)
# 
# Copyright (c) 2022 Project Nayuki. (MIT License)
# https://www.nayuki.io/page/aa-tree-set
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
# - The above copyright notice and this permission notice shall be included in
#   all copies or substantial portions of the Software.
# - The Software is provided "as is", without warranty of any kind, express or
#   implied, including but not limited to the warranties of merchantability,
#   fitness for a particular purpose and noninfringement. In no event shall the
#   authors or copyright holders be liable for any claim, damages or other
#   liability, whether in an action of contract, tort or otherwise, arising from,
#   out of or in connection with the Software or the use or other dealings in the
#   Software.
# 


class AaTreeSet:
	
	def __init__(self, coll=None):
		self.clear()
		if coll is not None:
			for val in coll:
				self.add(val)
	
	
	def clear(self):
		self.root = AaTreeSet.Node.EMPTY_LEAF
		self.size = 0
	
	
	def __len__(self):
		return self.size
	
	
	def __contains__(self, val):
		node = self.root
		while node is not AaTreeSet.Node.EMPTY_LEAF:
			if val < node.value:
				node = node.left
			elif val > node.value:
				node = node.right
			else:
				return True
		return False
	
	
	def add(self, val):
		if val in self:
			return
		self.root = self.root.add(val)
		self.size += 1
	
	
	def remove(self, val):
		self.root, found = self.root.remove(val)
		if not found:
			raise KeyError(str(val))
		self.size -= 1
	
	
	def discard(self, val):
		self.root, found = self.root.remove(val)
		if found:
			self.size -= 1
	
	
	# Note: Not fail-fast on concurrent modification.
	def __iter__(self):
		stack = []
		node = self.root
		while True:
			while node is not AaTreeSet.Node.EMPTY_LEAF:
				stack.append(node)
				node = node.left
			if len(stack) == 0:
				break
			node = stack.pop()
			yield node.value
			node = node.right
	
	
	# For unit tests
	def check_structure(self):
		visited = set()
		if self.root.check_structure(visited) != self.size or len(visited) != self.size:
			raise AssertionError()
	
	
	
	class Node:
		
		def __init__(self, val=None):
			self.value = val
			if val is None:  # For the singleton empty leaf node
				self.level = 0
			else:  # Normal non-leaf nodes
				self.level = 1
				self.left  = AaTreeSet.Node.EMPTY_LEAF
				self.right = AaTreeSet.Node.EMPTY_LEAF
		
		
		def add(self, val):
			if self is AaTreeSet.Node.EMPTY_LEAF:
				return AaTreeSet.Node(val)
			if val < self.value:
				self.left = self.left.add(val)
			elif val > self.value:
				self.right = self.right.add(val)
			else:
				raise ValueError("Value already in tree")
			return self._skew()._split()  # Rebalance this node
		
		
		def remove(self, val):
			EMPTY = AaTreeSet.Node.EMPTY_LEAF
			if self is EMPTY:
				return (EMPTY, False)
			
			if val < self.value:
				self.left, found = self.left.remove(val)
			elif val > self.value:
				self.right, found = self.right.remove(val)
			else:  # Remove value at this node
				found = True
				if self.left is not EMPTY:
					# Find predecessor node
					temp = self.left
					while temp.right is not EMPTY:
						temp = temp.right
					self.value = temp.value  # Replace value with predecessor
					self.left, fnd = self.left.remove(self.value)  # Remove predecessor node
					assert fnd
				elif self.right is not EMPTY:
					# Find successor node
					temp = self.right
					while temp.left is not EMPTY:
						temp = temp.left
					self.value = temp.value  # Replace value with successor
					self.right, fnd = self.right.remove(self.value)  # Remove successor node
					assert fnd
				else:
					assert self.level == 1
					return (EMPTY, True)
			
			# Rebalance this node if a child was lowered
			if not found or self.level == min(self.left.level, self.right.level) + 1:
				return (self, found)
			if self.right.level == self.level:
				self.right.level -= 1
			self.level -= 1
			result = self._skew()
			result.right = result.right._skew()
			if result.right.right is not EMPTY:
				result.right.right = result.right.right._skew()
			result = result._split()
			result.right = result.right._split()
			return (result, True)
		
		
		#       |          |
		#   A - B    ->    A - B
		#  / \   \        /   / \
		# 0   1   2      0   1   2
		def _skew(self):
			assert self is not AaTreeSet.Node.EMPTY_LEAF
			if self.left.level < self.level:
				return self
			result = self.left
			self.left = result.right
			result.right = self
			return result
		
		
		#   |                      |
		#   |                    - B -
		#   |                   /     \
		#   A - B - C    ->    A       C
		#  /   /   / \        / \     / \
		# 0   1   2   3      0   1   2   3
		def _split(self):
			assert self is not AaTreeSet.Node.EMPTY_LEAF
			# Must short-circuit because if right.level < self.level, then right.right might not exist
			if self.right.level < self.level or self.right.right.level < self.level:
				return self
			result = self.right
			self.right = result.left
			result.left = self
			result.level += 1
			return result
		
		
		# For unit tests, invokable by the outer class.
		def check_structure(self, visitednodes):
			if self is AaTreeSet.Node.EMPTY_LEAF:
				return 0
			if self in visitednodes:
				raise AssertionError()
			visitednodes.add(self)
			
			value = self.value
			level = self.level
			left  = self.left
			right = self.right
			if value is None or left is None or right is None:
				raise AssertionError()
			if not (level > 0 and level == left.level + 1 and level - right.level in (0, 1)):
				raise AssertionError()
			if level == right.level and level == right.right.level:  # Must short-circuit evaluate
				raise AssertionError()
			if left != AaTreeSet.Node.EMPTY_LEAF and not (left.value < value):
				raise AssertionError()
			if right != AaTreeSet.Node.EMPTY_LEAF and not (right.value > value):
				raise AssertionError()
			
			size = 1 + left.check_structure(visitednodes) + right.check_structure(visitednodes)
			if not (2**level - 1 <= size <= 3**level - 1):
				raise AssertionError()
			return size


# Static initializer. A bit of a hack, but more elegant than using None values as leaf nodes.
AaTreeSet.Node.EMPTY_LEAF = AaTreeSet.Node()
